File: compiler/compiler/src/main/java/org/robovm/compiler/namespace/RoboVmLocations.java
package org.robovm.compiler.namespace;

import org.robovm.compiler.Version;
import org.robovm.compiler.config.Config;

import java.io.File;

/**
 * Default locations for RoboVM entities such as cache/sdk etc
 * @author dkimitsa
 */
public final class RoboVmLocations {
    private RoboVmLocations() {}

    ///  just ~/
    public final static File UserHome = new File(System.getProperty("user.home"));

    /// e.g. ~/.robovm
    public final static File roboVmBaseDir = new File(UserHome, ".robovm");

    /// e.g. ~/.robovm-sdks
    public final static File roboVmSdksBaseDir = new File(UserHome, ".robovm-sdks");

    /// home for current RoboVM version, e.g. ~/.robovm-sdks/robovm-2.3.24
    public final static File roboVmSdkDir = new File(roboVmSdksBaseDir,"robovm-" + Version.getCompilerVersion());

    /// e.g. ~/.robovm/cache
    public final static File roboVmCacheDir = new File(roboVmBaseDir, "cache");

    /**
     * RoboVm SDK home.
     * points to $roboVmSdk (e.g. ~/.robovm-sdks/robovm-2.3.24)
     * or development compiler project location
     */
    public final static Config.Home roboVmHome;
    static {
        // check if running in dev environment
        Config.Home candidate = Config.Home.suggestDevHome();
        // production: not validated to allow SDK extraction routine to work with dev status
        if (candidate == null) candidate = new Config.Home(roboVmSdkDir, false);
        roboVmHome = candidate;
    }

    /**
     * e.g. ~/.robovm/${fileName}
     * @return location for fileName withing Home dir
     */
    public static File inHomeDir(String fileName) {
        return new File(roboVmBaseDir, fileName);
    }

    /**
     * e.g. project/ios-module/robovm-build
     * @return name for build-folder withing given module path
     */
    public static File buildDir(File moduleDir) {
        return new File(moduleDir, "robovm-build");
    }

    /**
     * e.g. project/ios-module/robovm-build/${subDir}
     * @return path for subDir withing build-folder in given moduleDir
     */
    public static File inBuildDir(File moduleDir, String subDir) {
        return new File(buildDir(moduleDir), subDir);
    }
}

File: plugins/idea/src/main/java/org/robovm/idea/RoboVmLocations.kt
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/gpl-2.0.html>.
 */
package org.robovm.idea

import org.robovm.compiler.Version
import org.robovm.compiler.config.Config
import java.io.File

object RoboVmLocations {
    /**
     * RoboVm home: not validated to allow SDK extraction routine to work with dev status
     * (and not unpack files when not required)
     */
    @JvmStatic
    val roboVmHome: Config.Home by lazy {
        Config.Home.suggestDevHome() ?: Config.Home(sdkHome, false)
    }

    /**
     * SDK home = ~/.robovm-sdks/robovm-$version
     */
    @JvmStatic
    val sdkHome: File by lazy {
        File(sdkHomeBase, "robovm-" + Version.getCompilerVersion())
    }

    /**
     * Location for SDKs
     * - SDK will be unpacked here;
     * - used when running in non-dev mode
     */
    @JvmStatic
    val sdkHomeBase: File by lazy {
        File(System.getProperty("user.home"), ".robovm-sdks")
    }


    /**
     * Location for RoboVM cache files
     */
    @JvmStatic
    val cacheDir: File by lazy {
        File(System.getProperty("user.home"), ".robovm/cache")
    }
}